src/silentpayments.rs

```diff
@@ -0,0 +1,546 @@
+
+use crate::ffi::{self, CPtr};
+use crate::{constants, Keypair, PublicKey, SecretKey, XOnlyPublicKey};
+use crate::Secp256k1;
+use crate::Verification;
+
+fn copy_to_ffi_pubkey(pubkey: &PublicKey) -> ffi::PublicKey {
+
+    unsafe {
+        // Get a pointer to the inner ffi::PublicKey
+        let ffi_pubkey_ptr: *const ffi::PublicKey = pubkey.as_c_ptr();
+        
+        // Dereference the pointer to get the ffi::PublicKey
+        // Then create a copy of it
+        (*ffi_pubkey_ptr).clone()
+    }
+}
+
+
+/// Struct to store recipient data
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct SilentpaymentsRecipient(ffi::SilentpaymentsRecipient);
+
+impl SilentpaymentsRecipient {
+
+    /// Get a new SilentpaymentsRecipient
+    pub fn new(scan_pubkey: &PublicKey,  spend_pubkey: &PublicKey, index: usize) -> Self {
+
+        Self(ffi::SilentpaymentsRecipient::new(
+            &copy_to_ffi_pubkey(scan_pubkey),
+            &copy_to_ffi_pubkey(spend_pubkey),
+            index
+        ))
+    }
+}
+
+/// Sender Output creation errors
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum SenderOutputCreationError {
+    /// Unexpected failures
+    Failure,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for SenderOutputCreationError {}
+
+impl fmt::Display for SenderOutputCreationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match self {
+            SenderOutputCreationError::Failure => write!(f, "Failed to create silent payments outputs"),
+        }
+    }
+}
+
+/// Create Silent Payment outputs for recipient(s).
+pub fn silentpayments_sender_create_outputs<C: Verification>(
+    secp: &Secp256k1<C>,
+    recipients: &mut [&SilentpaymentsRecipient],
+    smallest_outpoint: &[u8; 36],
+    taproot_seckeys: Option<&[&Keypair]>,
+    plain_seckeys: Option<&[&SecretKey]>,
+) -> Result<Vec<XOnlyPublicKey>, SenderOutputCreationError> {
+    let cx = secp.ctx().as_ptr();
+    let n_tx_outputs = recipients.len();
+
+    unsafe {
+        let mut out_pubkeys = vec![ffi::XOnlyPublicKey::new(); n_tx_outputs];
+        let mut out_pubkeys_ptrs: Vec<_> = out_pubkeys.iter_mut().map(|k| k as *mut _).collect();
+
+        for (i, pubkey) in out_pubkeys.iter_mut().enumerate() {
+            out_pubkeys_ptrs[i] = pubkey as *mut _;
+        }
+
+        let ffi_recipients_ptrs: &mut [*const ffi::SilentpaymentsRecipient] =
+                    transmute::<&mut [&SilentpaymentsRecipient], &mut [*const ffi::SilentpaymentsRecipient]>(recipients);
+
+        let (ffi_taproot_seckeys, n_taproot_seckeys) = match taproot_seckeys {
+            Some(keys) => {
+                let ffi_keys: &[*const ffi::Keypair] = transmute::<&[&Keypair], &[*const ffi::Keypair]>(taproot_seckeys.unwrap());
```

Comment:
Pointer casting is recommended over transmuting pointers and references, but both approaches are unsound if `Keypair` is not guaranteed to have the same layout as `ffi::Keypair`. Adding `#[repr(transparent)]` to `Keypair` fixes this.

Same comment for `SecretKey` a few lines down, and some more in this file, please check all uses of `transmute` and use pointer casting where possible, and review if the source and destination type are layout-compatible.

Reply:
Thanks. Done in ac90bf3093977945f19efbf2c6e088efb7b193a8 and 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e